airbyte-integrations/connectors/source-netsuite/source_netsuite/streams.py

```diff
@@ -0,0 +1,266 @@
+        resp = response.json()
+        has_more = resp.get("hasMore")
+        if has_more:
+            return {"offset": resp["offset"] + resp["count"]}
+        return None
+
+    def format_date(self, last_modified_date: str) -> str:
+        # the date format returned is differnet than what we need to use in the query
+        lmd_datetime = datetime.strptime(last_modified_date, self.output_datetime_format)
+        return lmd_datetime.strftime(self.input_datetime_format)
+
+    def request_params(self, next_page_token: Mapping[str, Any] = None, **kwargs) -> MutableMapping[str, Any]:
+        params = {}
+        if next_page_token:
+            params.update(**next_page_token)
+        return params
+
+    def fetch_record(self, record: Mapping[str, Any], request_kwargs: Mapping[str, Any]) -> Iterable[Mapping[str, Any]]:
+        url = record["links"][0]["href"]
+        args = {"method": "GET", "url": url, "params": {"expandSubResources": True}}
+        prep_req = self._session.prepare_request(requests.Request(**args))
+        response = self._send_request(prep_req, request_kwargs)
+        # sometimes response.status_code == 400,
+        # but contains json elements with error description,
+        # to avoid passing it as TYPE: RECORD, we filter response by status
+        if response.status_code == requests.codes.ok:
+            yield response.json()
+
+    def parse_response(
+        self,
+        response: requests.Response,
+        stream_state: Mapping[str, Any],
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+        **kwargs,
+    ) -> Iterable[Mapping]:
+
+        records = response.json().get("items")
+        request_kwargs = self.request_kwargs(stream_slice, next_page_token)
+        if records:
+            for record in records:
+                # make sub-requests for each record fetched
+                yield from self.fetch_record(record, request_kwargs)
+
+    def should_retry(self, response: requests.Response) -> bool:
+        if response.status_code in NETSUITE_ERRORS_MAPPING.keys():
+            message = response.json().get("o:errorDetails")
+            if isinstance(message, list):
+                error_code = message[0].get("o:errorCode")
+                known_error = NETSUITE_ERRORS_MAPPING.get(response.status_code)
+                if error_code in known_error.keys():
+                    setattr(self, "raise_on_http_errors", False)
+                    self.logger.warn(
+                        f"Stream `{self.name}`: {known_error.get(error_code)}, full error message: {message}",
+                    )
+                    pass
+                else:
+                    return super().should_retry(response)
+        return super().should_retry(response)
+
+
+class IncrementalNetsuiteStream(NetsuiteStream, ABC):
+    @property
+    def cursor_field(self) -> str:
+        return "lastModifiedDate"
+
+    def filter_records_newer_than_state(
```

Comment:
why do we need this if we pass `lastmodified` date in query params?

Reply:
Nice question! 

Since the stream_slices return the `1/1/2022` format of start/end dates, we cannot achieve data granularity based on `date-time`, only the date format could be passed with the query. Therefore, once we fetch data incrementally, we need to filter the records that already exist in the destination to avoid duplicates. I agree, that this is a kind of spike, but it gives the desired result.